lib/elixir/src/elixir_with.erl

```diff
@@ -70,37 +70,49 @@ translate(Meta, Args, S) ->
   {Parts, [{do, Expr} | ExprList]} = elixir_utils:split_last(Args),
   case ExprList of
     [{else, ElseExpr}] ->
-      {TCases, TS} = translate_case(Parts, {ok, Expr}, fun(X) -> {error, X} end, S),
-      translate_else(Meta, TCases, ElseExpr, TS);
+      {TCase, TS, HasMatch} = translate_case(Parts, {ok, Expr}, fun(X) -> {error, X} end, S),
+      translate_else(Meta, TCase, ElseExpr, TS, HasMatch);
     [] ->
-      translate_case(Parts, Expr, fun(X) -> X end, S)
+      {TCase, TS, _HasMatch} = translate_case(Parts, Expr, fun(X) -> X end, S),
+      {TCase, TS}
   end.
 
 translate_case(Parts, DoExpr, Wrapper, S) ->
-  Cases = build_case(Parts, DoExpr, Wrapper),
-  {TCases, TS} = elixir_translator:translate(Cases, S#elixir_scope{extra=nil}),
-  {TCases, elixir_scope:mergec(S, TS)}.
-
-translate_else(Meta, WithCases, ElseExpr, S) ->
+  {Case, HasMatch} = build_case(Parts, DoExpr, Wrapper, false),
+  {TCase, TS} = elixir_translator:translate(Case, S#elixir_scope{extra=nil}),
+  {TCase, elixir_scope:mergec(S, TS), HasMatch}.
+
+translate_else(Meta, WithCases, _ElseExpr, S, false) ->
+  Warning = "else clauses never match",
```

Comment:
Mmm, I don't really like that because few things in Elixir "have effect". Maybe "no else clauses will ever match" can be even better than my previous suggestion.